src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,672 @@
+	: device_t(mconfig, type, tag, owner, clock)
+	, m_brg(*this, "brg")
+	, m_txa_handler(*this)
+	, m_rts_handler(*this)
+	, m_cka_handler(*this)
+	, m_cts(0)
+	, m_dcd(0)
+	, m_irq(0)
+	, m_rts(0)
+	, m_bgr_divisor(0)
+	, m_divisor(0)
+	, m_id(id)
+	, m_ext(ext)
+{
+}
+
+
+void z180asci_channel_base::device_resolve_objects()
+{
+	// resolve callbacks
+	m_txa_handler.resolve_safe();
+	m_rts_handler.resolve_safe();
+	m_cka_handler.resolve_safe();
+}
+
+void z180asci_channel_base::device_start()
+{
+	save_item(NAME(m_asci_cntla));
+	save_item(NAME(m_asci_cntlb));
+	save_item(NAME(m_asci_stat));
+	save_item(NAME(m_asci_tdr));
+	save_item(NAME(m_asci_rdr));
+	if (m_ext)
+	{
+		save_item(NAME(m_asci_ext));
+		save_item(NAME(m_asci_tc.w));
+	}
+	save_item(NAME(m_tsr));
+	save_item(NAME(m_rsr));
+	save_item(NAME(m_data_fifo));
+	save_item(NAME(m_error_fifo));
+	save_item(NAME(m_fifo_wr));
+	save_item(NAME(m_fifo_rd));
+	save_item(NAME(m_rx_error));
+
+	save_item(NAME(m_cts));
+	save_item(NAME(m_dcd));
+	save_item(NAME(m_irq));
+	save_item(NAME(m_txa));
+	save_item(NAME(m_rxa));
+	save_item(NAME(m_rts));
+
+	save_item(NAME(m_bgr_divisor));
+	save_item(NAME(m_divisor));
+
+	save_item(NAME(m_clock_state));
+	save_item(NAME(m_tx_state));
+	save_item(NAME(m_tx_parity));
+	save_item(NAME(m_tx_bits));
+	save_item(NAME(m_tx_counter));
+	save_item(NAME(m_rx_state));
+	save_item(NAME(m_rx_parity));
+	save_item(NAME(m_rx_bits));
+	save_item(NAME(m_rx_counter));
+
+	save_item(NAME(m_rx_enabled));
+}
+
+void z180asci_channel_base::device_reset()
+{
+	m_asci_ext = 0;
+	m_asci_tc.w = 0;
+
+	m_fifo_wr = 0;
+	m_fifo_rd = 0;
+
+	output_txa(1);
+	output_rts(1);
+
+	m_tsr = 0;
+	m_rxa = 1;
+	m_clock_state = 0;
+	m_tx_state = STATE_STOP;
+	m_tx_parity = 0;
+	m_tx_bits = 0;
+	m_tx_counter = 0;
+	m_rx_state = STATE_START;
+	m_rx_parity = 0;
+	m_rx_bits = 0;
+	m_rx_counter = 0;
+	m_rx_enabled = true;
+}
+
+void z180asci_channel_base::device_add_mconfig(machine_config &config)
+{
+	CLOCK(config, m_brg, 0);
+	m_brg->signal_handler().append(*this, FUNC(z180asci_channel_base::brg_wr));
+}
+
+void z180asci_channel_base::device_clock_changed()
+{
+	 // Divide ratio
+	m_divisor = (m_asci_ext & Z180_ASEXT_X1_BIT_CLK0) ? 1 : ((m_asci_cntlb & Z180_CNTLB_DR) ? 64 : 16);
+
+	if ((m_asci_cntlb & Z180_CNTLB_SS) == Z180_CNTLB_SS)
+	{
+		// External clock
+		m_bgr_divisor = 0;
+	}
+	else
+	{
+		if (m_asci_ext & Z180_ASEXT_BRG0_MODE)
+		{
+			// Extended boud rate generator mode
+			m_bgr_divisor = m_asci_tc.w + 2;
+		}
+		else
+		{
+			// Regular bitrate generator mode
+			m_bgr_divisor = 1 << (m_asci_cntlb & Z180_CNTLB_SS);
+			m_bgr_divisor *= ((m_asci_cntlb & Z180_CNTLB_CTS_PS) ? 30 : 10); // Prescale
+		}
+	}
+
+	if (m_bgr_divisor)
+	{
+		LOG("Z180 ASCI%d set bitrate %d\n", m_id, uint32_t(clock() / m_bgr_divisor / m_divisor));
+		m_brg->set_clock(DERIVED_CLOCK(1,m_bgr_divisor));
```

Comment:
If you implement this as a clock running at the BRG rate, performance will be relatively poor due to the high timer frequency.  If you do it with a pair of timers, performance will be better, at the slight cost in accuracy of not being quantised to the BRG output frequency.

Also, you should check that you’ve spelled your member names correctly before auto-completing them everywhere.